src/compiler/checker.ts

```diff
@@ -338,7 +338,15 @@ module ts {
                 // Locals of a source file are not in scope (because they get merged into the global symbol table)
                 if (location.locals && !isGlobalSourceFile(location)) {
                     if (result = getSymbol(location.locals, name, meaning)) {
-                        break loop;
+                        // Type parameters of a function are in scope in the entire function declaration, including the parameter
+                        // list and return type. However, local types are only in scope in the function body.
+                        if (!(meaning & SymbolFlags.Type) ||
+                            !(result.flags & (SymbolFlags.Type & ~SymbolFlags.TypeParameter)) ||
+                            !isFunctionLike(location) ||
+                            lastLocation === (<FunctionLikeDeclaration>location).body) {

```

Comment:
Top-level block scoped declarations are always in the `locals` of the function itself. We need them to be in the same declaration space as parameters and top-level `var` and `function` declarations in order to properly detect collisions. I thought about having local types in a `locals` attached to the body, but it doesn't work when a symbol is declared both as a value and a type--you can't split one meaning off and put it in a different declaration space.